### PR TITLE
feat(resource-policy): delete manifests when policy value is delete

### LINKED ISSUE #3150: Resource policy flag ignores value
The annotation for keeping a persistent resource seems to not behave as expected.  Consider this brief chart:

```
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: {{ .Values.volume.name }}
  namespace: {{ .Values.namespace }}
  annotations:
    "volume.beta.kubernetes.io/storage-class": "{{ .Values.volume.name }}"
    "helm.sh/resource-policy": {{ .Values.volume.resourcePolicy }}
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: {{ .Values.volume.storage }}
```
The `"helm.sh/resource-policy"` annotation is what I'm focused on.

When the annotation is set to `keep`, I can install the chart and delete it properly--the volumeclaim is correctly kept, rather than being deleted with the rest of the release.

When I remove the annotation altogether, I can install the chart and delete it properly--the volume claim is correctly deleted with the rest of the release.

When the annotation is set to anything other than `keep` (e.g. `delete`, hard-coded or through .Values), I can install the chart properly, but the delete does not function as expected--the volumeclaim is not deleted with the rest of the release

I would expect that a value besides keep would not cause the volume to be kept after a delete.

### LINKED ISSUE #4713: Empty helm.sh/resource-policy annotation should not keep the resource
Output of `helm version`: v2.10.0

I was surprised that when using the stable/postgresql chart, the PVC was not destroyed when i delete the Helm release.

In this chart, the default value for "helm.sh/resource-policy" annotation is "" and not "keep" so the PVC should be destroyed.

After looking in pkg/tiller/resource_policy.go, seems that the "if resourcePolicyType == keepPolicy" should be complemented by an "else" that adds the manifest in the "remaining" array.

### LINKED ISSUE #5394: Resource delete policy differs between upgrades and deletes
As of https://github.com/helm/helm/pull/5225, [resource policy annotations](https://github.com/helm/helm/blob/33589472cf02d8d10e96e06085d95791c0611a75/docs/charts_tips_and_tricks.md#tell-tiller-not-to-delete-a-resource) are respected during upgrades. The logic for testing whether a resource should be deleted is implemented separately in the delete and upgrade cases, however, and leads to differing behaviors.

For example:

| Annotation                        | helm upgrade result | helm delete result |
|-----------------------------------|---------|--------|
| `"helm.sh/resource-policy": keep` | kept    | kept   |
| `"helm.sh/resource-policy": KEEP` | deleted  | kept   |
| `"helm.sh/resource-policy": ""`   | deleted  | kept   |

It seems like it would be best for upgrade behavior to match the more conservative and long-standing delete behavior.

### PR DESCRIPTION
**Pull request by jlegrone**
Closes #3150, closes #4713, closes #5394

**What this PR does / why we need it**:

This PR allows opting in to resource deletion if the `helm.sh/resource-policy` annotation has a value of `delete`. This PR also makes a small change related to https://github.com/helm/helm/pull/5225 in order to respond to resource policy annotation values consistently between upgrades and deletes (always keep resources if they have the annotation unless the value is `delete`).

From https://github.com/helm/helm/issues/3150#issuecomment-448436708:

> For more background, the particular issue I'm running into is I have charts with this annotation set, but there's no way to override the annotation when it comes time to install via --set parameters -- the resources are always retained. I've tried null, ~, delete, empty strings and so forth. The key always continues to be present.

A time when this might be convenient is if a chart is meant to ship with production-ready values by default, and therefore may opt to protect against data loss by using this annotation on resources such as a PerisistentVolumeClaim. The most common current approach is to support this configuration via a separate value, for example:

https://github.com/helm/charts/blob/cbd5e811a44c7bac6226b019f1d1810ef5ee45fa/stable/lamp/templates/pvc.yaml#L11-L13

However it is more flexible to be able to specify a default set of annotations in values.yaml, and allow end users (or parent charts) to override the value of those annotations if desired. This works well in any case where a chart allows specifying arbitrary annotations (eg. [kibana](https://github.com/helm/charts/blob/b3226cf40a681859092312306e5c48ea7cd9d05e/stable/kibana/values.yaml#L158)), but does not already specifically support the resource policy annotation.

**Special notes for your reviewer**:

Relates to https://github.com/helm/helm/pull/4715

**If applicable**:
- [x] this PR contains documentation
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

File: pkg/tiller/resource_policy.go
Comment (jlegrone) [inline pkg/tiller/resource_policy.go]: The manifest is now appended to the `keep` slice any time the resource policy annotation has been set (and its value is not a variant of `delete`).  This differs from the previous behavior, where `keep` was only updated when `resourcePolicyType == keepPolicy`.

It seems like the end behavior is the same, but I'm not sure if this could cause issues.

**Comment by jlegrone** [comment]
@technosophos would you care to take a look at this?

**Comment by technosophos** [comment]
I'll defer to @bacongobbler on this one.

File: pkg/kube/resource_policy.go
Comment (jlegrone) [inline pkg/kube/resource_policy.go]: I was on the fence about this, should we just remove `KeepPolicy` since we never actually need to reference it, or at least make these policy consts unexported?

**Comment by jlegrone** [comment]
@technosophos do you have time for a quick look at this, or to tag any other maintainers? I think now that this is sized as a large two approvals are required.

I've broken the commits out into functional changes, tests, and documentation, so hopefully it's now a little easier to review.

Most of the size is from tests, the main diff is here (27 lines): https://github.com/helm/helm/pull/5092/commits/55fbed95d586db554b33cfa4dceff1475d2d4d41. Thank you!

**Review by adamreese — APPROVED** [review]
Thanks for including the unit tests!

**Review by bacongobbler — APPROVED** [review]
Manually tested for a quick sanity check. LGTM!

**Pull request merged**
